src/main/java/teammates/logic/core/FeedbackSessionsLogic.java

```diff
@@ -235,17 +236,35 @@ public boolean isFeedbackSessionHasQuestionForStudents(
     /**
      * Checks whether a student has completed a feedback session.
      *
-     * <p> If there is no question for students, the feedback session is completed</p>
+     * <p>If feedback session consists of all team questions, session is completed by student only
+     * if someone from the team has responded. If feedback session has some individual questions,
+     * session is completed only if the student has responded to the individual questions
+     * (regardless of the completion status of the team questions).</p>
      */
-    public boolean isFeedbackSessionCompletedByStudent(FeedbackSessionAttributes fsa, String userEmail) {
-        if (frLogic.hasGiverRespondedForSession(userEmail, fsa.getFeedbackSessionName(), fsa.getCourseId())) {
+    public boolean isFeedbackSessionCompletedByStudent(FeedbackSessionAttributes fsa, String userEmail, String userTeam) {
+        String feedbackSessionName = fsa.getFeedbackSessionName();
+        String courseId = fsa.getCourseId();
+
+        List<FeedbackQuestionAttributes> allQuestions =
+                fqLogic.getFeedbackQuestionsForStudents(feedbackSessionName, courseId);
```

Comment:
This might actually incur many data read. @damithc can we evaluate the volume of such call?
Before this change, this execution flow only performs key only queries (in `frLogic.hasGiverRespondedForSession` and `fqLogic.sessionHasQuestions`). This solution is clean if adding back entity reads will not incur any concern on the cost.
From the logs, I don't see it though. Perhaps it is about the period.

Reply:
I have refactored to only perform key-only queries.